### DISCUSSION TITLE
what is a "Trackblock" (and do we need that)?

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#31] [Q&A] [answered]
this is in MackieC4
  
```
def trBlock(self, trackOffset, blocksize):

        block = []

        tracks = self.song().visible_tracks

        for track in range(0, blocksize):

            block.extend([str(tracks[(trackOffset + track)].name)])
````
I THINK that it's related to the bank of 8 concept of the MackieController, but am not sure.
Also, do we need/want that at all?
The only place where this gets called is in
`    
```
def trname_changestate(self, tid, track, r=0):

        if r == 1:

            pass

        else:

            self.trBlock(0, len(LiveUtils.getTracks()))
````
and when doing `refresh_state`
Reason for asking: It bugs me and sometimes throws errors when closing Live

**Comment by BilldarBagdar** [comment, marked as answer]
I would tend to agree, "track block" sounds like the 8 faders or channels on the Control Universal Pro and XT controller.  But the purpose isn't as clear to me.  I suspect one should be able to "page" through "track blocks" (8, 16, 24 tracks at a time however many controllers you have connected.)  The only way I suspect that could impact the C4 code would be somewhere you could navigate tracks by "pages" instead of just track-to-track-to-track one at a time.

If the code is throwing errors, then it is being called from somewhere.  Try experimenting by commenting out the body of the method(s) and see what happens.  Often the key to cleaning up a mess is to make another slightly bigger or different mess and analyze the differences and similarities between the two inputs and results.  

Or try to figure out if the method is ever called when r = 1 or what that 'r' is.  It looks like just a Boolean signal. So the method does the r = 0 case by default, but if r = 1, the method just "passes".  What happens if r = any other value? Same thing as r = 0, the default behavior;

If it is only throwing errors sometimes, can you tell what's going on when it does? Like what's different than last time when it didn't throw errors?

**Reply by markusschloesser** [reply]
the whole Mackie thing is really oldschool and with basically no framework support. And yes, it its built on the premise of 8 somethings, in this case encoders (and faders for MCU), which you can page. We however have a completely different paging mechanism.
As I am still messing around with the destroy and the error only happens when closing Live, it is low on my prio. But will check out your tips.

**Comment by markusschloesser** [comment]
for documentation's sake, this is the error:

```
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 722, in <lambda>
RemoteScriptError: cb = lambda: self.trname_changestate(tid, track, ret)
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 812, in trname_changestate
RemoteScriptError: self.trBlock(0, len(LiveUtils.getTracks()))
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 272, in trBlock
RemoteScriptError: block.extend([str(tracks[(trackOffset + track)].name)])  # MS IndexError when closing Live, again a Lambda thing!!
RemoteScriptError: IndexError
RemoteScriptError: Index out of range
```

**Comment by BilldarBagdar** [comment]
`cb = lambda: self.trname_changestate(tid, track, ret)`  A "lambda" is a "nameless function", so it is an anti-pattern to (give one a name by) assign one to a variable like`cb`.  But the pattern here looks like a way of assigning "callback functions" for the listeners to call.

The idea of a callback is central to the idea of "listeners".  It works in this way. The app that wants to be notified when some event occurs, registers with the app that "produces" the event.  The registration says in effect, "whenever the event I'm interested in "fires", call this method I've registered with you so I can do stuff in my own code related to that event firing"

The implementation here starts at the bottom of the file on line 812 where the actual callback function is defined `def trname_changestate(self, tid, track, r=0):` and then on line 721 where the `def add_trname_listener(self, tid, track, ret=0):` method is defined, that `trname_changestate` function is assigned as the "callback" within the listener framework.  and then up at line 272 is where the actual error is encountered, code execution actually fails.  

But I don't understand how the callback is necessarily related to this error. (There would seem to be a lot of "under the hood" wiring I can't see.)  The error itself appears to be a garden variety "index out of bounds" error, and could maybe be fixed with these changes?  
```
    def trBlock(self, trackOffset, blocksize):
        block = []
        tracks = self.song().visible_tracks
        for track_index in range(0, blocksize):
            if len(tracks) > trackOffset + track_index:
                trk_nme = tracks[(trackOffset + track_index)].name
                # not just adding the track name, adding a list with one element that is the track name
                block.extend([str(trk_nme)])
            else:
                block.extend(['fake Track NAME'])
``` 

But index OOB seems like more of a symptom, than a disease in this case... And since I don't have much of a clue about "listeners" it's just a super-wild-ass-guess (SWAG)

**Reply by markusschloesser** [reply]
I implemented this in the "implementLiveObjectValid" branch

**Comment by markusschloesser** [comment]
closed
